[TACHYON-534][TACHYON-543] Round Robin Allocation Policy + Allocator Test Framework

tianyin opened the pull request:
Sorry I still submit the two together as they are highly correlated...

calvinjia commented on servers/src/main/java/tachyon/worker/block/allocator/AllocatorFactory.java:
```diff
@@ -35,8 +35,10 @@ public static Allocator create(AllocatorType allocatorType, BlockMetadataManager
+        return new RoundRobinAllocator(metaManager);
       default:
-        return new GreedyAllocator(metaManager);
+        return new MaxFreeAllocator(metaManager);
```
No need to change the default in this PR?

tianyin replied:
I didn't change the default.

I think the rebased_worker_wip branch uses MaxFreeAllocator, while the
master branch uses GreedyAllocator.

Let me know if I'm wrong. I can definitely change back to Greedy.

~t

On 8 July 2015 at 12:21, Calvin Jia <notifications@github.com> wrote:

> In
> servers/src/main/java/tachyon/worker/block/allocator/AllocatorFactory.java
> <https://github.com/amplab/tachyon/pull/1040#discussion_r34187655>:
>
> >        default:
> > -        return new GreedyAllocator(metaManager);
> > +        return new MaxFreeAllocator(metaManager);
>
> No need to change the default in this PR?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/1040/files#r34187655>.
>



-- 
Tianyin XU,
http://cseweb.ucsd.edu/~tixu/

calvinjia replied:
Oh I see, let's keep the greedy version for now since thats the default in master.

tianyin replied:
I changed it back now (committed).

~t

On 8 July 2015 at 13:38, Calvin Jia <notifications@github.com> wrote:

> In
> servers/src/main/java/tachyon/worker/block/allocator/AllocatorFactory.java
> <https://github.com/amplab/tachyon/pull/1040#discussion_r34195503>:
>
> >        default:
> > -        return new GreedyAllocator(metaManager);
> > +        return new MaxFreeAllocator(metaManager);
>
> Oh I see, let's keep the greedy version for now since thats the default in
> master.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/1040/files#r34195503>.
>



-- 
Tianyin XU,
http://cseweb.ucsd.edu/~tixu/

calvinjia commented:
+1 LGTM
We may want to decide on a standard way for a developer to hook into a contract test. For example this PR does it automatically whereas #1042 does not.

apc999 commented on servers/src/main/java/tachyon/worker/block/allocator/AllocatorType.java:
```diff
@@ -33,4 +33,8 @@
    */
   GREEDY,
+  /**
+   * Round robin allocation to the storage dir
```
To be a little bit more clear, also describe if the round robin strategy is across tiers or within the same tier.

apc999 commented:
one javadoc issue left. I will LGTM this PR after it is fixed.

Thanks for the patch. It is really important.

tianyin commented:
added the doc.

~t

On 8 July 2015 at 16:56, Bin Fan <notifications@github.com> wrote:

> one javadoc issue left. I will LGTM this PR after it is fixed.
>
> Thanks for the patch. It is really important.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/1040#issuecomment-119762241>.
>



-- 
Tianyin XU,
http://cseweb.ucsd.edu/~tixu/

tianyin commented:
Hi, do you guys know what is "failed"? @@
I clicked the "Details" button but failed to find the failed things...

calvinjia commented:
Jenkins, test this please.

calvinjia commented:
@tianyin it looks like Jenkins had an issue and aborted your tests.

haoyuan commented:
Jenkins, test this please.

ooq commented:
I see that last build took 39min and there was a timeout error from the console. Is there a way to let Jenkins rebuild?

tianyin commented:
@ooq what is the cause of the error? How can I check it?

haoyuan commented:
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3599/

haoyuan commented:
Jenkins, test this please.

ooq commented:
@tianyin can you try making another commit (whatever you like) to this PR, it might trigger Jenkins to rebuild.

tianyin commented:
@ooq ok... @.@

tianyin commented:
"All checks have passed"
@ooq nice hack!

ooq commented:
@tianyin I wasn't sure it would work :)

carsonwang commented:
Hi @tianyin , I updated the max free allocator which changes the behavior a little. Can you please update the unit tests if necessary? Thanks a lot.

tianyin commented:
Hi, @carsonwang, you are the one who is supposed to update the unit test :), as the purpose of writing unit tests is to make sure the implementation is correct.

The unit test of MaxFreeAllocator is at
tachyon-servers/src/test/java/tachyon/worker/block/allocator/MaxFreeAllocatorTest.java

There are also a contract testing to test some general property of a valid Allocation class.

carsonwang commented:
I though the current tests will fail after #1066 was merged. If this merges cleanly, just leave it or remove it, I'll update the unit tests later.

uronce-cc commented:
@calvinjia EvictorTest.java functions as contract test for all types of evictors, it is a parameterized junit test with evictor type as parameter.

tianyin commented:
@uronce-cc I think Calvin's point is to standardize the practices of writing the contract tests. Right now, it's pretty free-style and are different across the components. @calvinjia

uronce-cc commented:
got the point.


> 在 2015年7月11日，下午4:50，Tianyin Xu <notifications@github.com> 写道：
> 
> @uronce-cc I think Calvin's point is to standardize the practices of writing the contract tests. Right now, it's pretty free-style and are different across the components. @calvinjia
> 
> ―
> Reply to this email directly or view it on GitHub.
>

haoyuan commented:
We can’t automatically merge this pull request.

tianyin commented:
What's the reason, @haoyuan ?

tianyin commented:
@haoyuan do you know how to check and resolve conflicts?

ooq commented:
@tianyin you should pull master and rebase/merge with it.

tianyin commented:
@ooq yes, this is what I did...

tianyin commented:
@ooq oh, you mean I should pull master right now and resolve the conflict. 

Could you tell me how to do that? I forked amplab/tachyon to tianyin/tachyon and branched from tianyin/tachyon master. How to pull from the original amplab/tachyon? Or did I do something completely wrong?

ooq commented:
@tianyin Yeah, you branch is 222 commits behind master.
You can set amplab/tachyon as the upstream and pull from upstream/master.
https://help.github.com/articles/syncing-a-fork/

ooq commented:
Also, you might want to rebase instead of merging, given the many changes since you last pull master..

tianyin commented:
Thank you so much, @ooq !! I just followed the github tutorial to do a rebase.

tianyin commented:
@ooq @haoyuan  Can it be merged now? Let me know if there's any other problem.

tianyin commented:
>.<

haoyuan commented:
@tianyin You can find more detailed info here: http://tachyon-project.org/master/Contributing-to-Tachyon.html

tianyin commented:
@haoyuan Thanks, I didn't anticipate so many changes :-(

ooq commented:
@tianyin Check the compilation error : https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/3682/

tianyin commented:
@ooq yes, I know..... Thanks a lot, Qifan! It's really a lot of changes related to the PR and I'm addressing it to work with the new codebase. Thank you so much for helping me :-) Big appreciation!!!

tianyin commented:
@ooq @haoyuan 
So sorry for the mess. I've adjusted the implementation to the new code base and passed the tests. Could you please review it and merge it?

There's nothing changed in terms of implementation, but only adapting the code changes since last week.

ooq commented on servers/src/main/java/tachyon/worker/block/allocator/RoundRobinAllocator.java:
```diff
@@ -0,0 +1,120 @@
+  private BlockMetadataManagerView mManagerView;
+
+  // We need to remember the last dir index for every storage tier
+  private Map<StorageTierView, Integer> mTierViewDirViews = new HashMap<StorageTierView, Integer>();
```
NIT: Something more informative? e.g. mTiersToLastDirs?

ooq commented on servers/src/main/java/tachyon/worker/block/allocator/RoundRobinAllocator.java:
```diff
@@ -0,0 +1,120 @@
+  private TempBlockMeta allocateBlock(long userId, long blockId, long blockSize,
+      BlockStoreLocation location) throws IOException {
+    if (location.equals(BlockStoreLocation.anyTier())) {
+      int tierAlias = 0; //always starting from the first tier
```
The way you use `tierAlias` here (you called `.get(tierAlias)` at line 76) is actually `tierIndex`. I would try avoid using `tierAlias` as it is confusing..

ooq commented on servers/src/main/java/tachyon/worker/block/allocator/RoundRobinAllocator.java:
```diff
@@ -0,0 +1,120 @@
+   * @param blockSize: the requested block size
+   * @return: the index of the dir if nonnegative; -1 if fail to find a dir
+   */
+  private int getNextDirViewInTierView(StorageTierView tierView, long blockSize) 
```
You should name it to emphasize that it finds an available view, not just a next view. Also you can leave out `view` in the method name. We are always assuming using views anyway.

ooq commented on servers/src/main/java/tachyon/worker/block/allocator/RoundRobinAllocator.java:
```diff
@@ -0,0 +1,120 @@
+    } else {
+      StorageTierView tierView = mManagerView.getTierView(location.tierAlias());
+      StorageDirView dirView = tierView.getDirView(location.dir());
+      if (dirView.getAvailableBytes() >= blockSize) {
```
Is it ever going to be true? My understanding is that `getNextDirViewInTierView` should have iterated through all dirViews.

tianyin replied:
What do you mean? @_@

Yes, getNextDirViewInTierView does iterate through all dirs (starting from
the last dir).





On 11 July 2015 at 16:28, Qifan Pu <notifications@github.com> wrote:

> In
> servers/src/main/java/tachyon/worker/block/allocator/RoundRobinAllocator.java
> <https://github.com/amplab/tachyon/pull/1040#discussion_r34417235>:
>
> > +          return tierView.getDirView(dirViewIndex).createTempBlockMeta(userId, blockId, blockSize);
> > +        } else { // we didn't find one in this tier, go to next tier
> > +          tierAlias ++;
> > +        }
> > +      }
> > +    } else if (location.equals(BlockStoreLocation.anyDirInTier(location.tierAlias()))) {
> > +      StorageTierView tierView = mManagerView.getTierView(location.tierAlias());
> > +      int dirViewIndex = getNextDirViewInTierView(tierView, blockSize);
> > +      if (dirViewIndex >= 0) {
> > +        mTierViewDirViews.put(tierView, dirViewIndex); // update
> > +        return tierView.getDirView(dirViewIndex).createTempBlockMeta(userId, blockId, blockSize);
> > +      }
> > +    } else {
> > +      StorageTierView tierView = mManagerView.getTierView(location.tierAlias());
> > +      StorageDirView dirView = tierView.getDirView(location.dir());
> > +      if (dirView.getAvailableBytes() >= blockSize) {
>
> Is it ever going to be true? My understanding is that
> getNextDirViewInTierView should have iterated through all dirViews.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/1040/files#r34417235>.
>



-- 
Tianyin XU,
http://cseweb.ucsd.edu/~tixu/

ooq replied:
nevermind

ooq commented:
@tianyin Thanks for the rebase. I left some comments for `RoundRobinAllocator.java`

ooq commented:
Thanks. I only reviewed `RoundRobinAllocator.java` given the many code-reviews already. It looks good to me.

tianyin commented:
@ooq thanks, Qifan. The other parts are test cases and do not really change (instead of refining the comments).

apc999 commented:
+1 LGTM

haoyuan commented:
We can’t automatically merge this pull request.

tianyin commented:
@haoyuan yes, it becomes stale again T_T I resolved the conflicts with the current upstream. Please merge it.

haoyuan commented:
Thanks @tianyin .

@calvinjia @apc999 could you please take another look?

calvinjia commented:
+1 LGTM

apc999 commented:
+1 LGTM

tianyin commented:
@carsonwang sorry I didn't understand your message in the first place. The PR is merged and the tests are changed to the current behavior. Feel free to revise. Thanks!
